File: server/app/services/task/tools/bash.py
"""
Bash tool for executing shell commands within a workspace.

Provides the agent with the ability to run shell commands on the local system.
Commands are confined to the task's workspace directory to ensure isolation.
Supports configurable timeout and returns stdout, stderr, and exit code.

Security:
- Path traversal outside workspace is blocked
- Access to .meta directory is blocked (system-managed files)
"""

import asyncio
import os
import re
from pathlib import Path
from typing import Any, Dict, Optional

from app.services.task.tools.base import Tool
from app.logger import logger


class BashTool(Tool):
    """
    Tool for executing shell commands via subprocess.

    Returns structured output containing stdout, stderr, and exit code.
    Commands are executed with a configurable timeout to prevent hanging.
    When a workspace is set, commands are executed with CWD set to the
    workspace directory and path traversal is blocked.
    """

    DEFAULT_TIMEOUT = 30000

    BLOCKED_PATTERNS = [
        r"\.meta",
        r"\.\./",
        r"\.\.\\",
    ]

    def __init__(self, workspace: Optional[Path] = None):
        """
        Initialize BashTool with optional workspace confinement.

        Args:
            workspace: If set, commands run with CWD=workspace and
                       path traversal outside workspace is blocked.
        """
        self._workspace = workspace

    @property
    def name(self) -> str:
        return "bash"

    @property
    def schema(self) -> Dict[str, Any]:
        workspace_hint = ""
        if self._workspace:
            workspace_hint = (
                f" All file operations must be within {self._workspace}. "
                "Do not access paths outside this directory."
            )
        return {
            "type": "function",
            "function": {
                "name": "bash",
                "description": (
                    "Execute a shell command. Use this tool to run commands, "
                    "manage files, and interact with the system." + workspace_hint
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "command": {
                            "type": "string",
                            "description": "The shell command to execute",
                        },
                        "timeout": {
                            "type": "integer",
                            "description": "Timeout in milliseconds (default: 30000)",
                            "default": self.DEFAULT_TIMEOUT,
                        },
                    },
                    "required": ["command"],
                },
            },
        }

    async def execute(self, **kwargs) -> str:
        """
        Execute a shell command and return the result.

        When workspace is set:
        - CWD is set to the workspace directory
        - Commands attempting to traverse outside workspace are blocked
        - Commands attempting to access .meta directory are blocked

        Args:
            command: The shell command to execute.
            timeout: Timeout in milliseconds (default: 30000).

        Returns:
            JSON string with stdout, stderr, and exit_code fields.
        """
        command = kwargs.get("command", "")
        timeout_ms = kwargs.get("timeout", self.DEFAULT_TIMEOUT)
        timeout_sec = timeout_ms / 1000.0

        if not command:
            return '{"stdout": "", "stderr": "Error: empty command", "exit_code": 1}'

        if self._workspace:
            blocked_reason = self._is_blocked_command(command)
            if blocked_reason:
                logger.warning(
                    f"BashTool blocked command: {command[:200]} - {blocked_reason}"
                )
                return (
                    f'{{"stdout": "", "stderr": "Error: {blocked_reason}", '
                    f'"exit_code": 1}}'
                )

        cwd = str(self._workspace) if self._workspace else None
        logger.info(
            f"BashTool executing: {command[:200]} "
            f"(timeout: {timeout_ms}ms, cwd: {cwd or 'default'})"
        )

        try:
            process = await asyncio.create_subprocess_shell(
                command,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=cwd,
            )

            try:
                stdout_bytes, stderr_bytes = await asyncio.wait_for(
                    process.communicate(), timeout=timeout_sec
                )
            except asyncio.TimeoutError:
                process.kill()
                await process.wait()
                logger.warning(f"BashTool timeout: {command[:100]}")
                return '{"stdout": "", "stderr": "Error: command timed out", "exit_code": -1}'

            stdout = stdout_bytes.decode("utf-8", errors="replace")
            stderr = stderr_bytes.decode("utf-8", errors="replace")
            exit_code = process.returncode if process.returncode is not None else -1

            result = (
                f'{{"stdout": {self._escape(stdout)}, '
                f'"stderr": {self._escape(stderr)}, '
                f'"exit_code": {exit_code}}}'
            )

            logger.info(
                f"BashTool result: exit_code={exit_code}, "
                f"stdout_len={len(stdout)}, stderr_len={len(stderr)}"
            )
            return result

        except Exception as e:
            logger.error(f"BashTool error: {str(e)}", exc_info=True)
            return f'{{"stdout": "", "stderr": "Error: {self._escape_str(str(e))}", "exit_code": 1}}'

    def _is_blocked_command(self, command: str) -> Optional[str]:
        """
        Check if a command should be blocked.

        Blocks:
        - Path traversal outside workspace
        - Access to .meta directory (system-managed files)

        Args:
            command: The shell command to check.

        Returns:
            Error message if blocked, None if allowed.
        """
        check_part = _strip_heredoc_content(command)

        if self._contains_meta_reference(check_part):
            return "access to .meta directory is not allowed"

        if self._is_path_traversal(check_part):
            return "command attempts to access paths outside workspace"

        return None

    def _contains_meta_reference(self, command: str) -> bool:
        """
        Check if command references .meta directory.

        Args:
            command: The command to check.

        Returns:
            True if .meta is referenced.
        """
        for pattern in self.BLOCKED_PATTERNS:
            if re.search(pattern, command):
                if ".meta" in command:
                    return True
        return False

    def _is_path_traversal(self, command: str) -> bool:
        """
        Detect obvious path traversal attempts in a command.

        Blocks commands that reference paths outside the workspace
        using absolute paths or parent directory references.

        Only checks the command portion, not heredoc content.
        Heredoc content (<< 'DELIM' ... DELIM) is data, not paths.

        This is a best-effort check; it cannot guarantee complete
        confinement but catches the most common patterns.

        Args:
            command: The shell command to check.

        Returns:
            True if path traversal is detected.
        """
        if not self._workspace:
            return False

        workspace_str = str(self._workspace)

        parts = command.split()
        for part in parts:
            if part.startswith("/") and not part.startswith(workspace_str):
                if not _is_safe_absolute_path(part, self._workspace):
                    return True
            if ".." in part:
                resolved = _safe_resolve(part, self._workspace)
                if resolved and not str(resolved).startswith(workspace_str):
                    return True
        return False

    @staticmethod
    def _escape(text: str) -> str:
        """Escape a string for safe JSON embedding."""
        escaped = text.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")
        return f'"{escaped}"'

    @staticmethod
    def _escape_str(text: str) -> str:
        """Escape a string for safe embedding inside a JSON string value."""
        return text.replace("\\", "\\\\").replace('"', '\\"').replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")


def _strip_heredoc_content(command: str) -> str:
    """
    Remove heredoc content from a command string.

    Heredoc syntax: << 'DELIM' ... DELIM  or  << "DELIM" ... DELIM  or  << DELIM ... DELIM
    The content between delimiters is data (HTML, code, etc.),
    not file paths, and should not be checked for path traversal.

    Args:
        command: The full command string.

    Returns:
        Command with heredoc content removed (only the command
        portion before the heredoc marker is retained).
    """
    pattern = r"<<-?\s*['\"]?(\w+)['\"]?"
    match = re.search(pattern, command)
    if not match:
        return command
    delimiter = match.group(1)
    heredoc_start = match.start()
    return command[:heredoc_start]


def _is_safe_absolute_path(path_str: str, workspace: Path) -> bool:
    """
    Check if an absolute path is a common system utility or within workspace.

    System commands like /bin/ls, /usr/bin/python are allowed.
    Only file paths outside workspace are blocked.

    Args:
        path_str: The absolute path to check.
        workspace: The workspace path.

    Returns:
        True if the path is considered safe.
    """
    safe_prefixes = (
        "/bin/", "/usr/bin/", "/usr/local/bin/",
        "/sbin/", "/usr/sbin/",
        "/opt/homebrew/",
    )
    for prefix in safe_prefixes:
        if path_str.startswith(prefix):
            return True
    return False


def _safe_resolve(path_str: str, workspace: Path) -> Optional[Path]:
    """
    Safely resolve a path relative to the workspace.

    Args:
        path_str: The path string (may be relative).
        workspace: The workspace path.

    Returns:
        Resolved path or None if resolution fails.
    """
    try:
        if os.path.isabs(path_str):
            return Path(path_str).resolve()
        return (workspace / path_str).resolve()
    except (OSError, ValueError):
        return None

File: server/internal/service/task/tools/bash.go
package tools

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"private-buddy-server/internal/service/llm"

	applogger "private-buddy-server/internal/logger"
)

// BashTool executes shell commands within a workspace.
//
// Provides the agent with the ability to run shell commands on the local system.
// Commands are confined to the task's workspace directory to ensure isolation.
// Supports configurable timeout and returns stdout, stderr, and exit code.
//
// Security:
//   - Path traversal outside workspace is blocked
//   - Access to .meta directory is blocked (system-managed files)
type BashTool struct {
	workspace string // Working directory for command execution
}

// NewBashTool creates a BashTool with the given workspace directory.
// If workspace is set, commands run with CWD=workspace and path traversal is blocked.
func NewBashTool(workspace string) *BashTool {
	return &BashTool{workspace: workspace}
}

func (b *BashTool) Name() string { return "bash" }

func (b *BashTool) Schema() llm.FunctionDefinition {
	workspaceHint := ""
	if b.workspace != "" {
		workspaceHint = fmt.Sprintf(" All file operations must be within %s. Do not access paths outside this directory.", b.workspace)
	}
	return llm.FunctionDefinition{
		Name:        "bash",
		Description: "Execute a shell command. Use this tool to run commands, manage files, and interact with the system." + workspaceHint,
		Parameters: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"command": map[string]interface{}{
					"type":        "string",
					"description": "The shell command to execute",
				},
				"timeout": map[string]interface{}{
					"type":        "integer",
					"description": "Timeout in milliseconds (default: 30000)",
					"default":     30000,
				},
			},
			"required": []string{"command"},
		},
	}
}

// BashResult holds the structured output of a bash command execution.
type BashResult struct {
	Stdout   string `json:"stdout"`
	Stderr   string `json:"stderr"`
	ExitCode int    `json:"exit_code"`
}

// Execute runs a bash command and returns structured output.
// Handles timeout, security checks, and returns JSON with stdout/stderr/exit_code.
func (b *BashTool) Execute(args map[string]interface{}) (string, error) {
	command, _ := args["command"].(string)
	timeoutMs := 30000
	if t, ok := args["timeout"].(float64); ok {
		timeoutMs = int(t)
	}

	if command == "" {
		return `{"stdout": "", "stderr": "Error: empty command", "exit_code": 1}`, nil
	}

	if b.workspace != "" {
		if blocked := b.isBlockedCommand(command); blocked != "" {
			cmdPreview := command
			if len(cmdPreview) > 200 {
				cmdPreview = cmdPreview[:200]
			}
			applogger.L.Warn("BashTool blocked command", "command", cmdPreview, "reason", blocked)
			return fmt.Sprintf(`{"stdout": "", "stderr": "Error: %s", "exit_code": 1}`, blocked), nil
		}
	}

	cmdPreview := command
	if len(cmdPreview) > 200 {
		cmdPreview = cmdPreview[:200]
	}
	applogger.L.Info("BashTool executing", "command", cmdPreview, "timeout_ms", timeoutMs)

	var cmd *exec.Cmd
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd", "/c", command)
	} else {
		cmd = exec.Command("bash", "-c", command)
	}
	if b.workspace != "" {
		cmd.Dir = b.workspace
	}

	timeout := time.Duration(timeoutMs) * time.Millisecond
	timer := time.NewTimer(timeout)
	defer timer.Stop()

	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Start(); err != nil {
		result, _ := json.Marshal(BashResult{Stderr: err.Error(), ExitCode: 1})
		return string(result), nil
	}

	done := make(chan error, 1)
	go func() { done <- cmd.Wait() }()

	select {
	case err := <-done:
		exitCode := 0
		if err != nil {
			exitCode = 1
			if exitErr, ok := err.(*exec.ExitError); ok {
				exitCode = exitErr.ExitCode()
			}
		}
		result, _ := json.Marshal(BashResult{
			Stdout:   stdout.String(),
			Stderr:   stderr.String(),
			ExitCode: exitCode,
		})
		return string(result), nil
	case <-timer.C:
		cmd.Process.Kill()
		result, _ := json.Marshal(BashResult{Stderr: "Error: command timed out", ExitCode: -1})
		return string(result), nil
	}
}

// isBlockedCommand checks if a command should be blocked for security reasons.
// Blocks access to .meta directory and path traversal outside workspace.
func (b *BashTool) isBlockedCommand(command string) string {
	checkPart := stripHeredocContent(command)

	if strings.Contains(checkPart, ".meta") {
		return "access to .meta directory is not allowed"
	}

	if b.isPathTraversal(checkPart) {
		return "command attempts to access paths outside workspace"
	}

	return ""
}

// isPathTraversal checks if a command attempts to access paths outside the workspace.
// Examines command parts for absolute paths and ".." traversal patterns.
func (b *BashTool) isPathTraversal(command string) bool {
	if b.workspace == "" {
		return false
	}
	parts := strings.Fields(command)
	for _, part := range parts {
		if strings.HasPrefix(part, "/") && !strings.HasPrefix(part, b.workspace) {
			if !isSafeAbsolutePath(part) {
				return true
			}
		}
		if strings.Contains(part, "..") {
			resolved := safeResolve(part, b.workspace)
			if resolved != "" && !strings.HasPrefix(resolved, b.workspace) {
				return true
			}
		}
	}
	return false
}

// safeResolve resolves a path relative to the workspace and returns the cleaned path.
func safeResolve(pathStr, workspace string) string {
	if filepath.IsAbs(pathStr) {
		return filepath.Clean(pathStr)
	}
	joined := filepath.Join(workspace, pathStr)
	return filepath.Clean(joined)
}

// stripHeredocContent removes heredoc content from a command for security checking.
// Heredoc content may contain paths that should not be checked for traversal.
func stripHeredocContent(command string) string {
	idx := strings.Index(command, "<<")
	if idx < 0 {
		return command
	}
	return command[:idx]
}

// isSafeAbsolutePath checks if an absolute path is a known safe system directory.
// Paths under /bin/, /usr/bin/, /usr/local/bin/, /sbin/, /usr/sbin/, /opt/homebrew/ are allowed.
func isSafeAbsolutePath(pathStr string) bool {
	safePrefixes := []string{"/bin/", "/usr/bin/", "/usr/local/bin/", "/sbin/", "/usr/sbin/", "/opt/homebrew/"}
	for _, prefix := range safePrefixes {
		if strings.HasPrefix(pathStr, prefix) {
			return true
		}
	}
	return false
}
